File: XWeather/iOS/ViewControllers/WeatherPvc.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;

using CoreAnimation;
using Foundation;
using UIKit;

using SettingsStudio;

using XWeather.Clients;
using XWeather.Unified;

namespace XWeather.iOS
{
	public partial class WeatherPvc : UIPageViewController
	{

		List<UITableViewController> Controllers = new List<UITableViewController> (3);


		public WeatherPvc (IntPtr handle) : base (handle) { }


		public override void ViewDidLoad ()
		{
			WuClient.Shared.UpdatedSelected += handleFirstUpdatedSelected;

			base.ViewDidLoad ();

			initToolbar ();

			initControllers ();
		}


		public override void ViewWillAppear (bool animated)
		{
			base.ViewWillAppear (animated);

			updateToolbarButtons (true);
		}


		public override void PrepareForSegue (UIStoryboardSegue segue, NSObject sender)
		{
			updateToolbarButtons (false);

			if (segue.Identifier.Equals ("locationsSegue", StringComparison.OrdinalIgnoreCase))
			{
				var current = ViewControllers.FirstOrDefault ();

				if (current != null)
					Analytics.TrackPageViewEnd (ViewControllers.FirstOrDefault (), WuClient.Shared.Selected);
			}
		}


		public override UIStatusBarStyle PreferredStatusBarStyle () => UIStatusBarStyle.LightContent;


		partial void closeClicked (NSObject sender)
		{
			updateToolbarButtons (true);

            DismissViewController(true, () =>
            {
                var current = ViewControllers.FirstOrDefault();

                if (current != null)
                    Analytics.TrackPageViewStart(current, childPageName(current), WuClient.Shared.Selected);
            });
		}


		partial void settingsClicked (NSObject sender)
		{
			var settingsString = UIApplication.OpenSettingsUrlString?.ToString ();

			if (!string.IsNullOrEmpty (settingsString))
			{
				var settingsUrl = NSUrl.FromString (settingsString);

				if (UIApplication.SharedApplication.OpenUrl (settingsUrl))
				{
					Analytics.TrackPageView (Pages.Settings.Name ());
				}
			}
		}


		void updateToolbarButtons (bool dismissing)
		{
			foreach (var button in toolbarButtons)
			{
				button.Hidden = dismissing ? button.Tag > 1 : button.Tag < 2;
			}

			pageIndicator.Hidden = !dismissing;
		}


		void handleFirstUpdatedSelected (object sender, EventArgs e)
		{
			WuClient.Shared.UpdatedSelected -= handleFirstUpdatedSelected;
			WuClient.Shared.UpdatedSelected += handleUpdatedSelected;

			refreshForUpdatedSelected ();
		}


		void handleUpdatedSelected (object sender, EventArgs e)
		{
			refreshForUpdatedSelected ();

			var current = ViewControllers.FirstOrDefault ();

			if (current != null)
				Analytics.TrackPageViewStart (current, childPageName (current), WuClient.Shared.Selected);
		}


		void refreshForUpdatedSelected ()
		{
			BeginInvokeOnMainThread (() =>
			{
				updateToolbarButtons (true);
				reloadData ();
			});
		}

		void reloadData ()
		{
			updateBackground ();

			foreach (var controller in Controllers) controller?.TableView?.ReloadData ();
		}


		void updateBackground ()
		{
			var location = WuClient.Shared.Selected;

			var random = location == null || Settings.RandomBackgrounds;

			var layer = View.Layer.Sublayers [0] as CAGradientLayer;

			if (layer == null)
			{
				layer = new CAGradientLayer ();
				layer.Frame = View.Bounds;
				View.Layer.InsertSublayer (layer, 0);
			}

			var gradients = location.GetTimeOfDayGradient (random);

			if (layer?.Colors?.Length > 0 && layer.Colors [0] == gradients.Item1 [0] && layer.Colors [1] == gradients.Item1 [1])
				return;

			CATransaction.Begin ();
			CATransaction.AnimationDuration = 1.5;
			layer.Colors = gradients.Item1;
			CATransaction.Commit ();
		}


		void initControllers ()
		{
			Controllers = new List<UITableViewController> { Storyboard.Instantiate<DailyTvc> (), Storyboard.Instantiate<HourlyTvc> (), Storyboard.Instantiate<DetailsTvc> () };

			GetNextViewController += (pvc, rvc) => rvc.Equals (Controllers [0]) ? Controllers [1] : rvc.Equals (Controllers [1]) ? Controllers [2] : null;

			GetPreviousViewController += (pvc, rvc) => rvc.Equals (Controllers [2]) ? Controllers [1] : rvc.Equals (Controllers [1]) ? Controllers [0] : null;

			WillTransition += (s, e) => { updateBackground (); };

			DidFinishAnimating += (s, e) =>
			{
				var index = Controllers.IndexOf ((UITableViewController)ViewControllers [0]);
				pageIndicator.CurrentPage = index;
				Settings.WeatherPage = index;
			};

			SetViewControllers (new [] { Controllers [Settings.WeatherPage] }, UIPageViewControllerNavigationDirection.Forward, false, (finished) => { getData (); });

			pageIndicator.CurrentPage = Settings.WeatherPage;

			updateBackground ();
		}


		void initToolbar ()
		{
			toolbarView.TranslatesAutoresizingMaskIntoConstraints = false;

			NavigationController.View.AddSubview (toolbarView);

			NavigationController.View.AddConstraints (NSLayoutConstraint.FromVisualFormat (@"H:|[toolbarView]|", 0, "toolbarView", toolbarView));
			NavigationController.View.AddConstraints (NSLayoutConstraint.FromVisualFormat (@"V:[toolbarView(44.0)]|", 0, "toolbarView", toolbarView));
		}


		Pages childPageName (UIViewController page)
		{
			if (page != null)
			{
				if (page.Equals (Controllers [0])) return Pages.WeatherDaily;
				if (page.Equals (Controllers [1])) return Pages.WeatherHourly;
				if (page.Equals (Controllers [2])) return Pages.WeatherDetails;
			}
			return Pages.Unknown;
		}

#if DEBUG

		void getData () => TestDataProvider.InitTestDataAsync ();
#else

		LocationProvider LocationProvider;

		void getData ()
		{
			if (LocationProvider == null) LocationProvider = new LocationProvider ();

			UIApplication.SharedApplication.NetworkActivityIndicatorVisible = true;

			System.Threading.Tasks.Task.Run (async () =>
			{
				var location = await LocationProvider.GetCurrentLocationCoordnatesAsync ();

				await WuClient.Shared.GetLocations (location);

				BeginInvokeOnMainThread (() =>
				{
					UIApplication.SharedApplication.NetworkActivityIndicatorVisible = false;
				});
			});
		}
#endif
	}
}

File: XWeather-Swift/XWeather/ViewControllers/WeatherPvc.swift
//
//  WeatherPvc.swift
//  XWeather
//
//  Created by Colby Williams on 6/11/17.
//  Copyright © 2017 Colby Williams. All rights reserved.
//

import UIKit

class WeatherPvc: UIPageViewController, UIPageViewControllerDataSource, UIPageViewControllerDelegate {

	@IBOutlet var emptyView: UIView!
	@IBOutlet var toolbarView: UIView!
	@IBOutlet weak var loadingContainerView: UIView!
	@IBOutlet weak var loadingIndicatorView: UIActivityIndicatorView!
	@IBOutlet weak var pageIndicator: UIPageControl!
	@IBOutlet var toolbarButtons: [UIButton]!
	
	var controllers = [UITableViewController]()
	
	
	override func viewDidLoad() {
        super.viewDidLoad()

        initToolbar()
		
		initControllers()
    }
	
	override func viewWillAppear(_ animated: Bool) {
		super.viewWillAppear(animated)
		
		updateToolbarButtons(true)
	}
	
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
	
	
	@IBAction func closeClicked(_ sender: Any) {
	}
	
	@IBAction func settingsClicked(_ sender: Any) {
	}
	
	
	func updateToolbarButtons (_ dismissing: Bool) {
		
		for button in toolbarButtons {
			button.isHidden = dismissing ? button.tag > 1 : button.tag < 2
		}
		
		pageIndicator.isHidden = !dismissing
	}
	
	
	func reloadData () {
		
		updateBackground()
		
		for controller in controllers {
			controller.tableView.reloadData()
		}
	}
	
	
	func updateBackground() {
		
		let location = WuClient.Shared.selected
		
		let random = location == nil || Settings.randomBackgrounds
		
		var layer = view.layer.sublayers?[0] as? CAGradientLayer
		
		if layer == nil {
			layer = CAGradientLayer()
			layer!.frame = view.bounds
			view.layer.insertSublayer(layer!, at: 0)
		}
		
		let gradients = location?.timeOfDayIndex (random);
		
		if let colors = gradients?.colors {
			
			if let oldColor0 = layer?.colors?[0], let oldColor1 = layer?.colors?[1] {
				if oldColor0 as! CGColor == colors[0] && oldColor1 as! CGColor == colors[1] {
					return
				}
			}
			
			CATransaction.begin()
			CATransaction.setAnimationDuration(1.5)
			layer?.colors = colors
			CATransaction.commit ()
		}
	}
	
	
	func initControllers() {
		
		if let dailyTvc = storyboard?.instantiateViewController(withIdentifier: "DailyTvc") as? UITableViewController,
			let hourlyTvc = storyboard?.instantiateViewController(withIdentifier: "HourlyTvc") as? UITableViewController,
			let detailsTvc = storyboard?.instantiateViewController(withIdentifier: "DetailsTvc") as? UITableViewController {
			
			controllers += [ dailyTvc, hourlyTvc, detailsTvc ]
		}
	}
	
	
	func initToolbar() {
		
		toolbarView.translatesAutoresizingMaskIntoConstraints = false
		
		navigationController?.view.addSubview(toolbarView)
		
		navigationController?.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "H:|[toolbarView]|", options: .directionLeadingToTrailing, metrics: [:], views: ["toolbarView": toolbarView]))
		navigationController?.view.addConstraints(NSLayoutConstraint.constraints(withVisualFormat: "V:[toolbarView(44.0)]|", options: .directionLeadingToTrailing, metrics: [:], views: ["toolbarView": toolbarView]))
	}
	
	
	
	// MARK: - UIPageViewControllerDataSource
	
	func pageViewController(_ pageViewController: UIPageViewController, viewControllerBefore viewController: UIViewController) -> UIViewController? {
		return pageViewController.isEqual(controllers[2]) ? controllers[1] : pageViewController.isEqual(controllers[1]) ? controllers[0] : nil
	}
	
	func pageViewController(_ pageViewController: UIPageViewController, viewControllerAfter viewController: UIViewController) -> UIViewController? {
		return pageViewController.isEqual(controllers[0]) ? controllers[1] : pageViewController.isEqual(controllers[1]) ? controllers[2] : nil
	}
	
	
	// MARK: - UIPageViewControllerDelegate
	
	func pageViewController(_ pageViewController: UIPageViewController, willTransitionTo pendingViewControllers: [UIViewController]) {
		updateBackground()
	}
	
	func pageViewController(_ pageViewController: UIPageViewController, didFinishAnimating finished: Bool, previousViewControllers: [UIViewController], transitionCompleted completed: Bool) {
		if let indexController = pageViewController.viewControllers?[0] as? UITableViewController, let index = controllers.index(of: indexController) {
			pageIndicator.currentPage = index
			Settings.weatherPage = index
		}
	}
}
